src/Storages/S3Queue/S3QueueSource.cpp

```diff
@@ -42,33 +43,93 @@ StorageS3QueueSource::S3QueueKeyWithInfo::S3QueueKeyWithInfo(
 StorageS3QueueSource::FileIterator::FileIterator(
     std::shared_ptr<S3QueueFilesMetadata> metadata_,
     std::unique_ptr<GlobIterator> glob_iterator_,
+    size_t current_shard_,
     std::atomic<bool> & shutdown_called_)
     : metadata(metadata_)
     , glob_iterator(std::move(glob_iterator_))
     , shutdown_called(shutdown_called_)
+    , log(&Poco::Logger::get("StorageS3QueueSource"))
+    , sharded_processing(metadata->isShardedProcessing())
+    , current_shard(current_shard_)
 {
+    if (sharded_processing)
+    {
+        for (size_t i = 0; i < metadata->getProcessingThreadsNum(); ++i)
+            sharded_keys.emplace(i, std::deque<KeyWithInfoPtr>{});
+    }
 }
 
-StorageS3QueueSource::KeyWithInfoPtr StorageS3QueueSource::FileIterator::next()
+StorageS3QueueSource::KeyWithInfoPtr StorageS3QueueSource::FileIterator::next(size_t idx)
 {
     while (!shutdown_called)
     {
-        KeyWithInfoPtr val = glob_iterator->next();
+        KeyWithInfoPtr val{nullptr};
+
+        {
+            std::unique_lock lk(sharded_keys_mutex, std::defer_lock);
+            if (sharded_processing)
+            {
+                /// To make sure order on keys in each shard in sharded_keys
+                /// we need to check sharded_keys and to next() under lock.
+                lk.lock();
+
+                auto & keys = sharded_keys.at(idx);
+                if (!keys.empty())
+                {
+                    val = keys.front();
+                    keys.pop_front();
+                }
+            }
+
+            if (!val)
+            {
+                val = glob_iterator->next();
+                if (val && sharded_processing)
+                {
+                    auto shard = metadata->getProcessingIdForPath(val->key);
```

Comment:
why is it stored in `variable` called `shard` when it's a processing id?

Reply:
Leftovers - initially I called this method `getProcessingShardForPath`, so it was `shard`, then renamed the method but not the variable.